File: modules/brl/native/gametarget.java

import java.io.*;
import java.nio.*;
import java.net.*;
import java.util.*;
import java.text.*;
import java.lang.reflect.*;

class BBGameEvent{
	static final int KeyDown=1;
	static final int KeyUp=2;
	static final int KeyChar=3;
	static final int MouseDown=4;
	static final int MouseUp=5;
	static final int MouseMove=6;
	static final int TouchDown=7;
	static final int TouchUp=8;
	static final int TouchMove=9;
	static final int MotionAccel=10;
}

class BBGameDelegate{
	void StartGame(){}
	void SuspendGame(){}
	void ResumeGame(){}
	void UpdateGame(){}
	void RenderGame(){}
	void KeyEvent( int event,int data ){}
	void MouseEvent( int event,int data,float x,float y ){}
	void TouchEvent( int event,int data,float x,float y ){}
	void MotionEvent( int event,int data,float x,float y,float z ){}
	void DiscardGraphics(){}
}

class BBDisplayMode{
	public int width;
	public int height;
	public int format;
	public int hertz;
	public int flags;
	public BBDisplayMode(){}
	public BBDisplayMode( int width,int height ){ this.width=width;this.height=height; }
}

abstract class BBGame{

	protected static BBGame _game;

	protected BBGameDelegate _delegate;
	protected boolean _keyboardEnabled;
	protected int _updateRate;
	protected boolean _debugExs;
	protected boolean _started;
	protected boolean _suspended;
	protected long _startms;
	
	public BBGame(){
		_game=this;
		_debugExs=MonkeyConfig.CONFIG.equals( "debug" );
		_startms=System.currentTimeMillis();
	}
	
	public static BBGame Game(){
		return _game;
	}
	
	public boolean Started(){
		return _started;
	}
	
	public boolean Suspended(){
		return _suspended;
	}

	public void SetDelegate( BBGameDelegate delegate ){
		_delegate=delegate;
	}
	
	public BBGameDelegate Delegate(){
		return _delegate;
	}
	
	public void SetKeyboardEnabled( boolean enabled){
		_keyboardEnabled=enabled;
	}
	
	public boolean KeyboardEnabled(){
		return _keyboardEnabled;
	}

	public void SetUpdateRate( int hertz ){
		_updateRate=hertz;
	}
	
	public int UpdateRate(){
		return _updateRate;
	}

	public int Millisecs(){
		return (int)(System.currentTimeMillis()-_startms);
	}
	
	public void GetDate( int[] date ){
		int n=date.length;
		if( n>0 ){
			Calendar c=Calendar.getInstance();
			date[0]=c.get( Calendar.YEAR );
			if( n>1 ){
				date[1]=c.get( Calendar.MONTH )+1;
				if( n>2 ){
					date[2]=c.get( Calendar.DATE );
					if( n>3 ){
						date[3]=c.get( Calendar.HOUR_OF_DAY );
						if( n>4 ){
							date[4]=c.get( Calendar.MINUTE );
							if( n>5 ){
								date[5]=c.get( Calendar.SECOND );
								if( n>6 ){
									date[6]=c.get( Calendar.MILLISECOND );
								}
							}
						}
					}
				}
			}
		}
	}
	
	public int SaveState( String state ){
		return -1;
	}

	public String LoadState(){
		return "";
	}

	public String LoadString( String path ){
		byte[] bytes=LoadData( path );
		if( bytes!=null ) return loadString( bytes );
		return "";
	}
	
	public int CountJoysticks( boolean update ){
		return 0;
	}
	
	public boolean PollJoystick( int port,float[] joyx,float[] joyy,float[] joyz,boolean[] buttons ){
		return false;
	}
	
	public void SetMouseVisible( boolean visible ){
	}

	int GetDeviceWidth(){
		return 0;
	}
	
	int GetDeviceHeight(){
		return 0;
	}
	
	void SetDeviceWindow( int width,int height,int flags ){
	}
	
	BBDisplayMode[] GetDisplayModes(){
		return new BBDisplayMode[0];
	}
	
	BBDisplayMode GetDesktopMode(){
		return null;
	}
	
	void SetSwapInterval( int interval ){
	}
	
	public void OpenUrl( String url ){	
	}
	
	String PathToFilePath( String path ){
		return "";
	}
	
	//***** Java Game *****
	
	public RandomAccessFile OpenFile( String path,String mode ){
		try{
			return new RandomAccessFile( PathToFilePath( path ),mode );
		}catch( IOException ex ){
		}
		return null;
	}
	
	public InputStream OpenInputStream( String path ){
		try{
			if( path.startsWith( "http:" ) || path.startsWith( "https:" ) ){
				URL url=new URL( path );
				URLConnection con=url.openConnection();
				return new BufferedInputStream( con.getInputStream() );
			}
			return new FileInputStream( PathToFilePath( path ) );
		}catch( IOException ex ){
		}
		return null;
	}
	
	byte[] LoadData( String path ){
		try{
			InputStream in=OpenInputStream( path );
			if( in==null ) return null;
			
			ByteArrayOutputStream out=new ByteArrayOutputStream(1024);
			byte[] buf=new byte[4096];
			
			for(;;){
				int n=in.read( buf );
				if( n<0 ) break;
				out.write( buf,0,n );
			}
			in.close();
			return out.toByteArray();
			
		}catch( IOException e ){
		}
		return null;
	}
	
	//***** INTERNAL *****
	
	public void Quit(){
		System.exit( 0 );
	}
	
	public void Die( RuntimeException ex ){
	
		String msg=ex.getMessage();
		if( msg!=null && msg.equals("") ){
			Quit();
			return;
		}
		if( _debugExs ){
			if( msg==null ) msg=ex.toString();
			bb_std_lang.print( "Monkey Runtime Error : "+msg );
			bb_std_lang.print( bb_std_lang.stackTrace() );
		}
		throw ex;
	}

	void StartGame(){
	
		if( _started ) return;
		_started=true;
		
		try{
			synchronized( _delegate ){
				_delegate.StartGame();
			}
		}catch( RuntimeException ex ){
			Die( ex );
		}
	}
	
	void SuspendGame(){
	
		if( !_started || _suspended ) return;
		_suspended=true;
		
		try{
			synchronized( _delegate ){
				_delegate.SuspendGame();
			}
		}catch( RuntimeException ex ){
			Die( ex );
		}
	}
	
	void ResumeGame(){

		if( !_started || !_suspended ) return;
		_suspended=false;
		
		try{
			synchronized( _delegate ){
				_delegate.ResumeGame();
			}
		}catch( RuntimeException ex ){
			Die( ex );
		}
	}
	
	void UpdateGame(){

		if( !_started || _suspended ) return;

		try{
			synchronized( _delegate ){
				_delegate.UpdateGame();
			}
		}catch( RuntimeException ex ){
			Die( ex );
		}
	}
	
	void RenderGame(){
	
		if( !_started ) return;
		
		try{
			synchronized( _delegate ){
				_delegate.RenderGame();
			}
		}catch( RuntimeException ex ){
			Die( ex );
		}
	}
	
	void KeyEvent( int event,int data ){

		if( !_started ) return;
	
		try{
			synchronized( _delegate ){
				_delegate.KeyEvent( event,data );
			}
		}catch( RuntimeException ex ){
			Die( ex );
		}
	}
	
	void MouseEvent( int event,int data,float x,float y ){

		if( !_started ) return;
		
		try{
			synchronized( _delegate ){
				_delegate.MouseEvent( event,data,x,y );
			}
		}catch( RuntimeException ex ){
			Die( ex );
		}
	}
	
	void TouchEvent( int event,int data,float x,float y ){

		if( !_started ) return;
		
		try{
			synchronized( _delegate ){
				_delegate.TouchEvent( event,data,x,y );
			}
		}catch( RuntimeException ex ){
			Die( ex );
		}
	}
	
	void MotionEvent( int event,int data,float x,float y,float z ){

		if( !_started ) return;
		
		try{
			synchronized( _delegate ){
				_delegate.MotionEvent( event,data,x,y,z );
			}
		}catch( RuntimeException ex ){
			Die( ex );
		}
	}
	
	void DiscardGraphics(){

		if( !_started ) return;
		
		try{
			synchronized( _delegate ){
				_delegate.DiscardGraphics();
			}
		}catch( RuntimeException ex ){
			Die( ex );
		}
	}
	
	//***** Private *****
	
	private String toString( byte[] buf ){
		int n=buf.length;
		char tmp[]=new char[n];
		for( int i=0;i<n;++i ){
			tmp[i]=(char)(buf[i] & 0xff);
		}
		return new String( tmp );
	}
	
	private String loadString( byte[] buf ){
	
		int n=buf.length;
		StringBuilder out=new StringBuilder();
		
		int t0=n>0 ? buf[0] & 0xff : -1;
		int t1=n>1 ? buf[1] & 0xff : -1;
		
		if( t0==0xfe && t1==0xff ){
			int i=2;
			while( i<n-1 ){
				int x=buf[i++] & 0xff;
				int y=buf[i++] & 0xff;
				out.append( (char)((x<<8)|y) ); 
			}
		}else if( t0==0xff && t1==0xfe ){
			int i=2;
			while( i<n-1 ){
				int x=buf[i++] & 0xff;
				int y=buf[i++] & 0xff;
				out.append( (char)((y<<8)|x) ); 
			}
		}else{
			int t2=n>2 ? buf[2] & 0xff : -1;
			int i=(t0==0xef && t1==0xbb && t2==0xbf) ? 3 : 0;
			boolean fail=false;
			while( i<n ){
				int c=buf[i++] & 0xff;
				if( (c & 0x80)!=0 ){
					if( (c & 0xe0)==0xc0 ){
						if( i>=n || (buf[i] & 0xc0)!=0x80 ){
							fail=true;
							break;
						}
						c=((c & 0x1f)<<6) | (buf[i] & 0x3f);
						i+=1;
					}else if( (c & 0xf0)==0xe0 ){
						if( i+1>=n || (buf[i] & 0xc0)!=0x80 || (buf[i+1] & 0xc0)!=0x80 ){
							fail=true;
							break;
						}
						c=((c & 0x0f)<<12) | ((buf[i] & 0x3f)<<6) | (buf[i+1] & 0x3f);
						i+=2;
					}else{
						fail=true;
						break;
					}
				}
				out.append( (char)c );
			}
			if( fail ){
				return toString( buf );
			}
		}
		return out.toString();
	}
}

File: modules/brl/native/gametarget.cs

public class BBGameEvent{
	public const int None=0;
	public const int KeyDown=1;
	public const int KeyUp=2;
	public const int KeyChar=3;
	public const int MouseDown=4;
	public const int MouseUp=5;
	public const int MouseMove=6;
	public const int TouchDown=7;
	public const int TouchUp=8;
	public const int TouchMove=9;
	public const int MotionAccel=10;
}

public class BBGameDelegate{
	public virtual void StartGame(){}
	public virtual void SuspendGame(){}
	public virtual void ResumeGame(){}
	public virtual void UpdateGame(){}
	public virtual void RenderGame(){}
	public virtual void KeyEvent( int ev,int data ){}
	public virtual void MouseEvent( int ev,int data,float x,float y ){}
	public virtual void TouchEvent( int ev,int data,float x,float y ){}
	public virtual void MotionEvent( int ev,int data,float x,float y,float z ){}
	public virtual void DiscardGraphics(){}
}

public class BBDisplayMode{
	public int width;
	public int height;
	public int format;
	public int hertz;
	public int flags;
	
	public BBDisplayMode( int width=0,int height=0,int format=0,int hertz=0,int flags=0 ){
		this.width=width;
		this.height=height;
		this.format=format;
		this.hertz=hertz;
		this.flags=flags;
	}
}

public class BBGame{

	protected static BBGame _game;
	
	protected BBGameDelegate _delegate;
	protected bool _keyboardEnabled;
	protected int _updateRate;
	protected bool _debugExs;
	protected bool _started;
	protected bool _suspended;
	protected Stopwatch _stopwatch;
	
	public BBGame(){
		_game=this;
		_debugExs=(MonkeyConfig.CONFIG=="debug");
		_stopwatch=Stopwatch.StartNew();
		
	}
	
	public static BBGame Game(){
		return _game;
	}
	
	public virtual void SetDelegate( BBGameDelegate delegate_ ){
		_delegate=delegate_;
	}
	
	public BBGameDelegate Delegate(){
		return _delegate;
	}
	
	public virtual void SetKeyboardEnabled( bool enabled ){
		_keyboardEnabled=enabled;
	}
	
	public virtual void SetUpdateRate( int hertz ){
		_updateRate=hertz;
	}
	
	public virtual int Millisecs(){
		return (int)_stopwatch.ElapsedMilliseconds;
	}
	
	public virtual void GetDate( int[] date ){
		int n=date.Length;
		if( n>0 ){
			DateTime t=DateTime.Now;
			date[0]=t.Year;
			if( n>1 ){
				date[1]=t.Month;
				if( n>2 ){
					date[2]=t.Day;
					if( n>3 ){
						date[3]=t.Hour;
						if( n>4 ){
							date[4]=t.Minute;
							if( n>5 ){
								date[5]=t.Second;
								if( n>6 ){
									date[6]=t.Millisecond;
								}
							}
						}
					}
				}
			}
		}
	}
	
	public virtual String CurrentDate(){
		return DateTime.Now.ToString( "dd MMM yyyy" );
	}
	
	public virtual String CurrentTime(){
		return DateTime.Now.ToString( "HH\\:mm\\:ss" );
	}
	
	public virtual int SaveState( String state ){
		return -1;
	}
	
	public virtual String LoadState(){
		return "";
	}
	
	public virtual String LoadString( String path ){

		Stream stream=OpenInputStream( path );
		if( stream==null ) return "";
		
		StreamReader reader=new StreamReader( stream );
		String text=reader.ReadToEnd();
		reader.Close();

		return text;
	}
	
	public virtual int CountJoysticks( bool update ){
		return 0;
	}
	
	public virtual bool PollJoystick( int port,float[] joyx,float[] joyy,float[] joyz,bool[] buttons ){
		return false;
	}
	
	public virtual void OpenUrl( String url ){
	}
	
	public virtual void SetMouseVisible( bool visible ){
	}
	
	public virtual int GetDeviceWidth(){
		return 0;
	}
	
	public virtual int GetDeviceHeight(){
		return 0;
	}
	
	public virtual void SetDeviceWindow( int width,int height,int flags ){
	}
	
	public virtual BBDisplayMode[] GetDisplayModes(){
		return new BBDisplayMode[0];
	}
	
	public virtual BBDisplayMode GetDesktopMode(){
		return null;
	}
	
	public virtual void SetSwapInterval( int interval ){
	}
	
	public virtual String PathToFilePath( String path ){
		return "";
	}
	
	//***** C# extensions *****
	
	public virtual FileStream OpenFile( String path,FileMode mode ){
		try{
			return new FileStream( PathToFilePath( path ),mode );
		}catch( Exception ex ){
		}
		return null;
	}
	
	public virtual Stream OpenInputStream( String path ){
		return OpenFile( path,FileMode.Open );
	}
	
	public virtual byte[] LoadData( String path ){
	
		Stream stream=OpenInputStream( path );
		if( stream==null ) return null;
		
		//fixme: stream may not have a Length (will always for now though).
		//
		int len=(int)stream.Length;
		byte[] buf=new byte[len];
		int n=stream.Read( buf,0,len );
		stream.Close();
		if( n==len ) return buf;
	
		return null;
	}

	//***** INTERNAL *****
	
	public virtual void Quit(){
		_delegate=new BBGameDelegate();
	}
	
	public virtual bool Die( Exception ex ){
	
		if( ex.Message=="" ){
			Quit();
			return false;
		}
		if( _debugExs ){
			bb_std_lang.Print( "Monkey Runtime Error : "+ex.Message );
			bb_std_lang.Print( bb_std_lang.StackTrace() );
		}
		return true;
	}
	
	public virtual void StartGame(){
	
		if( _started ) return;
		_started=true;
		
		try{
			_delegate.StartGame();
		}catch( Exception ex ){
			if( Die( ex ) ) throw;
		}
	}
	
	public virtual void SuspendGame(){
	
		if( !_started || _suspended ) return;
		_suspended=true;
		
		try{
			_delegate.SuspendGame();
		}catch( Exception ex ){
			if( Die( ex ) ) throw;
		}
	}
	
	public virtual void ResumeGame(){
	
		if( !_started || !_suspended ) return;
		_suspended=false;
		
		try{
			_delegate.ResumeGame();
		}catch( Exception ex ){
			if( Die( ex ) ) throw;
		}
	}
	
	public virtual void UpdateGame(){
	
		if( !_started || _suspended ) return;
		
		try{
			_delegate.UpdateGame();
		}catch( Exception ex ){
			if( Die( ex ) ) throw;
		}
	}
	
	public virtual void RenderGame(){
	
		if( !_started ) return;
		
		try{
			_delegate.RenderGame();
		}catch( Exception ex ){
			if( Die( ex ) ) throw;
		}
	}
	
	public virtual void KeyEvent( int ev,int data ){

		if( !_started ) return;
		
		try{
			_delegate.KeyEvent( ev,data );
		}catch( Exception ex ){
			if( Die( ex ) ) throw;
		}
	}
	
	public virtual void MouseEvent( int ev,int data,float x,float y ){

		if( !_started ) return;
		
		try{
			_delegate.MouseEvent( ev,data,x,y );
		}catch( Exception ex ){
			if( Die( ex ) ) throw;
		}
	}
	
	public virtual void TouchEvent( int ev,int data,float x,float y ){

		if( !_started ) return;
		
		try{
			_delegate.TouchEvent( ev,data,x,y );
		}catch( Exception ex ){
			if( Die( ex ) ) throw;
		}
	}
	
	public virtual void MotionEvent( int ev,int data,float x,float y,float z ){

		if( !_started ) return;
		
		try{
			_delegate.MotionEvent( ev,data,x,y,z );
		}catch( Exception ex ){
			if( Die( ex ) ) throw;
		}
	}
}
